src/daft-functions/src/temporal/mod.rs

```diff
@@ -0,0 +1,174 @@
+use serde::{Deserialize, Serialize};
+
+#[cfg(feature = "python")]
+pub fn register_modules(parent: &Bound<PyModule>) -> PyResult<()> {
+    parent.add_function(wrap_pyfunction_bound!(py_dt_date::dt_date, parent)?)?;
+    parent.add_function(wrap_pyfunction_bound!(py_dt_day::dt_day, parent)?)?;
+    parent.add_function(wrap_pyfunction_bound!(
+        py_dt_day_of_week::dt_day_of_week,
+        parent
+    )?)?;
+    parent.add_function(wrap_pyfunction_bound!(py_dt_hour::dt_hour, parent)?)?;
+    parent.add_function(wrap_pyfunction_bound!(py_dt_minute::dt_minute, parent)?)?;
+    parent.add_function(wrap_pyfunction_bound!(py_dt_month::dt_month, parent)?)?;
+    parent.add_function(wrap_pyfunction_bound!(py_dt_second::dt_second, parent)?)?;
+    parent.add_function(wrap_pyfunction_bound!(py_dt_time::dt_time, parent)?)?;
+    parent.add_function(wrap_pyfunction_bound!(py_dt_year::dt_year, parent)?)?;
+    parent.add_function(wrap_pyfunction_bound!(truncate::py_dt_truncate, parent)?)?;
+    Ok(())
+}
+
+macro_rules! impl_temporal {
+    ($name_str:expr, $name:ident, $dt:ident, $py_name:ident, $dtype:ident) => {
+        #[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
+        pub struct $name {}
+
+        #[typetag::serde]
+        impl ScalarUDF for $name {
+            fn as_any(&self) -> &dyn std::any::Any {
+                self
+            }
+
+            fn name(&self) -> &'static str {
+                $name_str
+            }
+
+            fn to_field(&self, inputs: &[ExprRef], schema: &Schema) -> DaftResult<Field> {
+                match inputs {
+                    [input] => match input.to_field(schema) {
+                        Ok(field) if field.dtype.is_temporal() => {
+                            Ok(Field::new(field.name, DataType::$dtype))
+                        }
+                        Ok(field) => Err(DaftError::TypeError(format!(
+                            "Expected input to {} to be temporal, got {}",
+                            self.name(),
+                            field.dtype
+                        ))),
+                        Err(e) => Err(e),
+                    },
+                    _ => Err(DaftError::SchemaMismatch(format!(
+                        "Expected 1 input arg, got {}",
+                        inputs.len()
+                    ))),
+                }
+            }
+
+            fn evaluate(&self, inputs: &[Series]) -> DaftResult<Series> {
+                match inputs {
+                    [input] => input.$dt(),
+                    _ => Err(DaftError::ValueError(format!(
+                        "Expected 1 input arg, got {}",
+                        inputs.len()
+                    ))),
+                }
+            }
+        }
+
+        pub fn $dt(input: ExprRef) -> ExprRef {
+            ScalarFunction::new($name {}, vec![input]).into()
+        }
+
+        // We gotta throw these in a module to avoid name conflicts
```

Comment:
Nit: I wonder if there's a way to throw them all into the *same* `mod`, so we don't have so many Python modules flying around (1 for each function)

Or another idea might just be to have the macro prefix `py_$dt` to avoid the name conflict?

(Prefacing this with: I'm not the most familiar with macros and pyo3 semantics here)

Reply:
> Or another idea might just be to have the macro prefix py_$dt to avoid the name conflict?

yes we can do that by using the [paste](https://crates.io/crates/paste) crate. Normal macros don't allow for the `py_$dt` syntax.